crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,654 @@
+
+        let b = b.clone();
+
+        all_w.push(std::thread::spawn(move || {
+            for i in 0..SCOUNT {
+                b.wait();
+
+                PutInfo {
+                    addr,
+                    agent_seed: &agent_seed,
+                    test_prop: &format!("{i}"),
+                    ..Default::default()
+                }
+                .call()
+                .unwrap();
+            }
+        }));
+    }
+
+    for j in all_w {
+        j.join().unwrap();
+    }
+
+    w_done.store(true, std::sync::atomic::Ordering::SeqCst);
+
+    for j in all_r {
+        j.join().unwrap();
+    }
+
+    let addr = format!("http://{:?}/bootstrap/{}", addr, S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+    let res = res.into_iter().map(|m| m.test_prop).collect::<Vec<_>>();
+
+    assert_eq!(
+        &[
+            "7", "7", "7", "7", "7", "7", "7", "7", "7", "7", "7", "7", "7",
+            "7", "7", "7"
+        ],
+        res.as_slice()
+    );
+
+    println!("multi_thread_stress in {}s", start.elapsed().as_secs_f64());
+}
+
+#[test]
+fn expiration_prune() {
+    let s = BootstrapSrv::new(Config {
+        prune_interval: std::time::Duration::from_millis(5),
+        ..Config::testing()
+    })
+    .unwrap();
+    let addr = s.listen_addr();
+    let addr = format!("http://{:?}/bootstrap/{}", addr, S1);
+
+    // -- the entry that WILL get pruned -- //
+
+    let created_at = crate::now();
+    let expires_at =
+        created_at + std::time::Duration::from_millis(500).as_micros() as i64;
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    // -- the entry that WILL NOT get pruned -- //
+
+    let created_at = crate::now();
+    let expires_at =
+        created_at + std::time::Duration::from_secs(60).as_micros() as i64;
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        agent_seed: K2,
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    /*
+     * NOTE - we might be tempted to check that both entries got in there:
+     *
+     * ```
+     * let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+     * let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+     * assert_eq!(2, res.len());
+     * ```
+     *
+     * but the PUTs didn't error, and we have other tests for that.
+     *
+     * Windows is slow enough that the get will be flaky... sometimes the
+     * pruner will have already run on it.
+     */
+
+    std::thread::sleep(std::time::Duration::from_secs(1));
+
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+
+    assert_eq!(1, res.len());
+}
```

Comment:
do we want a test for InvalidPathSegment error?

Reply:
I'm not sure there *is* an invalid SpaceId. One app can even claim the empty one : ) You can get a base64 decode error. Do you think we need a test for that?

The agent key will similarly generate base64 decode errors, or signature validation errors if it is not a valid pub key.

Reply:
Oh, I see you mean the error generated if they don't even specify the right count of path segments. Yes I can test that.

Reply:
Done: https://github.com/holochain/kitsune2/pull/11/commits/8cfe688c6779c30e6b214b4bbcfb08eaa3da4652